Implemented IsSubType Module Successfully

Smit1603 opened the pull request:
- This pull request will resolve issue #18 
- I have implemented IsSubType Module and ran it succesfully
- Kindly review it and please let me know whether any improvement or changes are required or not
- @vgvassilev and @sudo-panda

vgvassilev commented on lib/Interpreter/InterOp.cpp:
```diff
@@ -2184,5 +2184,40 @@ namespace InterOp {
+    auto derivedType = clang::QualType::getFromOpaquePtr(subclass);
+    auto baseType = clang::QualType::getFromOpaquePtr(superclass);
+
+    if (derivedType.isNull() || baseType.isNull())
+    {
+      return false;
+    }
```
```suggestion
    if (derivedType.isNull() || baseType.isNull())
      return false;
```

vgvassilev commented on lib/Interpreter/InterOp.cpp:
```diff
@@ -2184,5 +2184,40 @@ namespace InterOp {
+    auto derivedDecl = derivedType->getAsCXXRecordDecl();
+    auto baseDecl = baseType->getAsCXXRecordDecl();
+
+    if (!derivedDecl || !baseDecl)
+    {
+      return false;
+    }
```
```suggestion
    if (!derivedDecl || !baseDecl)
      return false;
```

vgvassilev commented on lib/Interpreter/InterOp.cpp:
```diff
@@ -2184,5 +2184,40 @@ namespace InterOp {
+      return false;
+    }
+
+    if (derivedDecl == baseDecl)
+    {
+      return true;
+    }
```
```suggestion
    if (derivedDecl == baseDecl)
      return true;
```

vgvassilev commented on lib/Interpreter/InterOp.cpp:
```diff
@@ -2184,5 +2184,40 @@ namespace InterOp {
+    for (auto it = derivedDecl->bases_begin(); it != derivedDecl->bases_end(); it++)
+    {
+      auto baseRecordType = it->getType()->getAs<clang::RecordType>();
+      if (baseRecordType && baseRecordType->getDecl() == baseDecl)
+      {
+        return true;
+      }
```
```suggestion
      if (baseRecordType && baseRecordType->getDecl() == baseDecl)
        return true;
```

Smit1603 commented:
- @vgvassilev I have made changes as mentioned by you 
- Kindly review and pls let me know whether any more changes/improvements are required or not

sudo-panda commented on lib/Interpreter/InterOp.cpp:
```diff
@@ -2184,5 +2184,31 @@ namespace InterOp {
+
+  bool IsSubType(TCppType_t subclass, TCppType_t superclass)
+  {
+    auto derivedType = clang::QualType::getFromOpaquePtr(subclass);
+    auto baseType = clang::QualType::getFromOpaquePtr(superclass);
```
```suggestion
    auto derivedType = clang::QualType::getFromOpaquePtr(derived);
    auto baseType = clang::QualType::getFromOpaquePtr(base);
```

sudo-panda commented on unittests/InterOp/TypeReflectionTest.cpp:
```diff
@@ -346,9 +346,9 @@ TEST(TypeReflectionTest, DISABLED_IsSubType) {
-  // EXPECT_TRUE(InterOp::IsSubType(type_B, type_A));
-  // EXPECT_FALSE(InterOp::IsSubType(type_A, type_B));
-  // EXPECT_FALSE(InterOp::IsSubType(type_C, type_A));
+  EXPECT_TRUE(InterOp::IsSubType(type_B, type_A));
+  EXPECT_FALSE(InterOp::IsSubType(type_A, type_B));
+  EXPECT_FALSE(InterOp::IsSubType(type_C, type_A));
```
Can you add more test cases?

Smit1603 commented:
- @sudo-panda I have changed variables name as mentioned by you for consistency purpose
- With regards to extra test cases I tried to perform multilevel inheritance and tried above implemented module and found that test gives incorrect output 
- So I am currently resolving that issue and will commit everything after resolving above issues 
- Just making sure and informing you whether I am going in correct direction???

Smit1603 commented:
- @sudo-panda as mentioned above
- I have added few more test cases which can check for various types of inheritance
- I have also changed variable names for consistency purpose
- I hope this modified code will meet every test cases and its requirements
- Pls review and let me know whether any further changes/improvements are required or not

vgvassilev commented on lib/Interpreter/InterOp.cpp:
```diff
@@ -2184,5 +2184,35 @@ namespace InterOp {
+  {
+    auto derivedType = clang::QualType::getFromOpaquePtr(derived);
+    auto baseType = clang::QualType::getFromOpaquePtr(base);
+
```
I think we could implement that using `Sema::isDerivedFrom`: https://clang.llvm.org/doxygen/classclang_1_1Sema.html#ae9fa2ea2b4ce733c9f50241d25c45132

vgvassilev replied:
@sudo-panda, what should that function return if we ask a nested type if it is a subtype?

sudo-panda replied:
The input can be assumed to be a `RecordType`

sudo-panda replied:
Yeah I agree we are looking for something like "IsTypeDerivedFrom". 

I didn't know that inner classes are also referred to as subtypes.

vgvassilev replied:
See my initial comment, this is what we need to do, and change the interface name as @sudo-panda suggested.

vgvassilev replied:
@Smit1603 this is not yet done

Smit1603 commented:
- @vgvassilev and @sudo-panda pls review and let me know whether any further improvements/changes are required or not

vgvassilev commented on lib/Interpreter/InterOp.cpp:
```diff
@@ -2184,5 +2184,25 @@ namespace InterOp {
     return dims;
   }
+
+  bool IsTypeDerivedFrom(TCppType_t derived, TCppType_t base)
+  {
+    auto derivedType = clang::QualType::getFromOpaquePtr(derived);
+    auto baseType = clang::QualType::getFromOpaquePtr(base);
+
+    if (derivedType.isNull() || baseType.isNull())
+      return false;
+
+    auto derivedDecl = derivedType->getAsCXXRecordDecl();
+    auto baseDecl = baseType->getAsCXXRecordDecl();
+
+    if (!derivedDecl || !baseDecl)
+      return false;
+
+    if (derivedDecl == baseDecl)
+      return true;
+
+    return derivedDecl->isDerivedFrom(baseDecl);
+  }
```
```suggestion
  bool IsTypeDerivedFrom(SemaType *S, TCppType_t derived, TCppType_t base)
  {
    auto derivedType = clang::QualType::getFromOpaquePtr(derived);
    auto baseType = clang::QualType::getFromOpaquePtr(base);
    return S->isDerivedFrom(derivedType, baseType);
  }
```

I was suggesting something like that.

Smit1603 commented:
- @vgvassilev sry for misconception on my side
- I have modified code as mentioned by you
- I hope the module is correctly implemented now
- Pls review and let me know whether any further improvements are required or not

vgvassilev commented:
> * @vgvassilev sry for misconception on my side
> 
>     * I have modified code as mentioned by you
> 
>     * I hope the module is correctly implemented now
> 
>     * Pls review and let me know whether any further improvements are required or not

Please write a good commit message and make this PR with a single commit.

Smit1603 commented:
- @vgvassilev and @sudo-panda I have squashed all commit as asked
- Pls review and let me know whether is it done or not

vgvassilev commented:
As a commit message I can suggest: "Add an API checking if a type is derived from another."

Smit1603 commented:
- @vgvassilev I hope this works

vgvassilev commented on include/clang/Interpreter/InterOp.h:
```diff
@@ -161,6 +161,8 @@ namespace InterOp {
 
   TCppType_t GetTypeFromScope(TCppScope_t klass);
 
+  bool IsTypeDerivedFrom(TCppSema_t sema,TCppType_t derived, TCppType_t base);
```
```suggestion
  /// Check if a C++ type derives from another.
  bool IsTypeDerivedFrom(TCppSema_t sema, TCppType_t derived, TCppType_t base);
```

vgvassilev commented on lib/Interpreter/InterOp.cpp:
```diff
@@ -2184,5 +2184,15 @@ namespace InterOp {
     return dims;
   }
+
+  bool IsTypeDerivedFrom(TCppSema_t sema,TCppType_t derived, TCppType_t base)
```
```suggestion
  bool IsTypeDerivedFrom(TCppSema_t sema, TCppType_t derived, TCppType_t base)
```

Smit1603 replied:
Have performed this minor changes

vgvassilev reviewed (approved):
LGTM assuming the comments are addressed.
